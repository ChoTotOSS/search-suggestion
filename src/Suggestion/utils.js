const getClosestScrollParent = findParent.bind(
  null,
  node => node.scrollHeight > node.clientHeight
);

function findParent(finder, node, rootNode) {
  if (node !== null && node !== rootNode.parentNode) {
    if (finder(node)) {
      return node;
    } else {
      return findParent(finder, node.parentNode, rootNode);
    }
  } else {
    return null;
  }
}

export function scrollIntoView(node, rootNode) {
  const scrollParent = getClosestScrollParent(node, rootNode);
  if (scrollParent === null) {
    return;
  }

  // getComputedStyle is a function belong to "window"
  const scrollParentStyles = getComputedStyle(scrollParent);
  const scrollParentRect = scrollParent.getBoundingClientRect();
  const scrollParentBorderTopWidth = parseInt(
    scrollParentStyles.borderTopWidth,
    10
  );
  const scrollParentBorderBottomWidth = parseInt(
    scrollParentStyles.borderBottomWidth,
    10
  );
  const scrollParentTop = scrollParentRect.top + scrollParentBorderTopWidth;
  const nodeRect = node.getBoundingClientRect();
  const nodeOffsetTop = nodeRect.top + scrollParent.scrollTop;
  const nodeTop = nodeOffsetTop - scrollParentTop;
  if (nodeTop < scrollParent.scrollTop) {
    // the item is above the scrollable area
    scrollParent.scrollTop = nodeTop;
  } else if (
    nodeTop +
      nodeRect.height +
      scrollParentBorderTopWidth +
      scrollParentBorderBottomWidth >
    scrollParent.scrollTop + scrollParentRect.height
  ) {
    // the item is below the scrollable area
    scrollParent.scrollTop =
      nodeTop +
      nodeRect.height -
      scrollParentRect.height +
      scrollParentBorderTopWidth +
      scrollParentBorderBottomWidth;
  }
  // the item is within the scrollable area (do nothing)
}

export function getNodeById(id) {
  return document.getElementById(id);
}
