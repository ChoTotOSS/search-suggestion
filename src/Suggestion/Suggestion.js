import React from 'react';
import PropTypes from 'prop-types';
import { scrollIntoView, getNodeById } from './utils';

const KEYS_EVENT = {
  UP: 38,
  DOWN: 40,
  ESCAPE: 27,
  ENTER: 13
};

class Suggestion extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      inputValue: '',
      dataLength: 0,
      currentValue: null,
      selectedItem: {},
      isOutsideClick: false,
      isOpen: false,
      highlightedIndex: 0
    };
  }

  componentDidMount() {
    document.addEventListener('mousedown', this.clickOutside);
    if (this.props.initInputValue) {
      this.setState({
        inputValue: this.props.initInputValue,
        dataLength: this.props.items.length
      });
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.items.length !== this.props.items.length) {
      this.setState({
        dataLength: nextProps.items.length
      });
    }
    if (nextProps.initInputValue !== this.props.initInputValue) {
      this.setState({
        inputValue: nextProps.initInputValue || ''
      });
    }
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.clickOutside);
  }

  onChange = e => {
    this.setState({
      inputValue: e.target.value,
      isOpen: true
    });
  };

  onBlurInput = e => {
    if (this.state.isOutsideClick) {
      this.setState({
        isOpen: false
      });
    }
  };

  onFocusInput = e => {
    this.setState({
      isOutsideClick: false,
      isOpen: true
    });
  };

  onSelectItem = item => {
    let displayName = {};
    if (this.props.getDisplayName) {
      displayName = {
        inputValue: this.props.getDisplayName(item) || ''
      };
    }

    if (this.props.onSelectedItem) {
      this.props.onSelectedItem(item);
    }

    this.setState({
      selectedItem: item,
      isOpen: false,
      ...displayName
    });
  };

  onKeyDown = e => {
    let itemIndex = this.state.highlightedIndex;
    if (this.state.isOpen && !this.state.isOutsideClick) {
      switch (e.keyCode) {
        case KEYS_EVENT.UP:
          {
            itemIndex--;
            if (itemIndex < 0) {
              itemIndex = 0;
            }
          }
          break;
        case KEYS_EVENT.DOWN:
          {
            itemIndex++;
            if (itemIndex + 1 > this.state.dataLength) {
              itemIndex = this.state.dataLength - 1;
            }
          }
          break;
        case KEYS_EVENT.ENTER:
          {
            this.onSelectItem(this.props.items[itemIndex]);
          }
          break;
        default:
          itemIndex = -1;
          break;
      }

      const item = getNodeById(`item_${itemIndex}`);
      if (!item) return;
      const container = getNodeById('__listItemProps');
      scrollIntoView(item, container);

      this.setState({
        highlightedIndex: itemIndex
      });
    }
  };

  getInputProps = props => {
    const {
      onChange: handleChange,
      onBlur: handleBlur,
      onFocus: handleFocus,
      placeholder
    } = props;
    const { inputValue } = this.state;

    const onKeyDown = e => {
      this.onKeyDown(e);
    };

    const onChange = e => {
      this.onChange(e);
      if (typeof handleChange === 'function') {
        handleChange(e.target.value);
      }
    };

    const onBlur = e => {
      this.onBlurInput(e);
      if (typeof handleBlur === 'function') {
        handleBlur(e.target.value);
      }
    };

    const onFocus = e => {
      this.onFocusInput(e);
      if (typeof handleFocus === 'function') {
        handleFocus(e.target.value);
      }
    };

    return {
      id: '__inputItemProps',
      placeholder,
      onKeyDown,
      onChange,
      onBlur,
      onFocus,
      value: inputValue
    };
  };

  getListItemProps = props => {
    return {
      id: '__listItemProps'
    };
  };

  getItemProps = props => {
    const { item, index, handleSelectItem } = props;
    const onClick = e => {
      this.onSelectItem(item);
      if (typeof handleSelectItem === 'function') {
        handleSelectItem(e.target.value);
      }
    };

    return {
      onClick,
      id: `item_${index}`
    };
  };

  clickOutside = e => {
    if (this.__autoComplete && !this.__autoComplete.contains(e.target)) {
      this.setState({
        isOutsideClick: true,
        isOpen: false
      });
    }
  };

  clearInputValue = () => {
    this.setState({
      inputValue: ''
    });
  };

  render() {
    const { inputValue, selectedItem, isOpen, highlightedIndex } = this.state;
    const { items, id = 'autoComplete' } = this.props;
    return (
      <div id={id} ref={el => (this.__autoComplete = el)}>
        {this.props.children({
          getInputProps: this.getInputProps,
          getItemProps: this.getItemProps,
          getListItemProps: this.getListItemProps,
          clearInputValue: this.clearInputValue,
          inputValue,
          selectedItem,
          isOpen,
          items,
          highlightedIndex
        })}
      </div>
    );
  }
}

Suggestion.propTypes = {
  children: PropTypes.func.isRequired,
  items: PropTypes.array.isRequired,
  initInputValue: PropTypes.string,
  placeholder: PropTypes.string,
  id: PropTypes.string,
  onSelectedItem: PropTypes.func,
  onChange: PropTypes.func,
  onBlur: PropTypes.func,
  onFocus: PropTypes.func,
  handleSelectItem: PropTypes.func,
  getDisplayName: PropTypes.func
};

Suggestion.defaultProps = {
  items: []
};

export default Suggestion;
