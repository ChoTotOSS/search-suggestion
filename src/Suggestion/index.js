import Suggestion from './Suggestion';

export default Suggestion;
