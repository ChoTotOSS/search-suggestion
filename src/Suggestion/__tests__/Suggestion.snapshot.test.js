import React from 'react';
import renderer from 'react-test-renderer';
import Suggestion from '../';

const { describe, it } = global;

describe('Suggestion render list item', () => {
  it('should render list items', () => {
    const tree = renderer
      .create(
        <Suggestion>
          {({ getItemProps }) => (
            <div>
              <span {...getItemProps({ index: 0, item: 0 })}>0</span>
              <span {...getItemProps({ index: 1, item: 1 })}>1</span>
              <span {...getItemProps({ index: 2, item: 2 })}>2</span>
              <span {...getItemProps({ index: 3, item: 3 })}>3</span>
              <span {...getItemProps({ index: 4, item: 4 })}>4</span>
            </div>
          )}
        </Suggestion>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
