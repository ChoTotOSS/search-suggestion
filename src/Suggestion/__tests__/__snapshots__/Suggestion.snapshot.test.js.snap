// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Suggestion render list item should render list items 1`] = `
<div
  id="autoComplete"
>
  <div>
    <span
      id="item_0"
      onClick={[Function]}
    >
      0
    </span>
    <span
      id="item_1"
      onClick={[Function]}
    >
      1
    </span>
    <span
      id="item_2"
      onClick={[Function]}
    >
      2
    </span>
    <span
      id="item_3"
      onClick={[Function]}
    >
      3
    </span>
    <span
      id="item_4"
      onClick={[Function]}
    >
      4
    </span>
  </div>
</div>
`;
